### ISSUE TITLE
Run periodic job on OpenShift CI

### ISSUE THREAD

**Issue by amitkrout** [openshift/odo#2931]
/kind feature

## Which functionality do you think we should add?
Currently we have high volume of test running for each pr which consumes almost 3hrs. So the plan here is to distribute the test in different types run. For example in nightly run all the test script will run periodically on master branch over a period of time which we should not bother about. The pr test will be validated against integration test only. 

Also there is one more benefit to add the nightly run on master branch is to catch the flake if there is any to avoid unpleasant experience while running pr test.

**Comment by kadel** [comment]
/remove-area release-eng
/triage needs-informion

What kind of jobs or test will run nightly?

**Comment by amitkrout** [comment]
> /remove-area release-eng
> /remove-kind feature
> /kind test
> /triage needs-information
> 
> What kind of jobs or test will run nightly?

Updated the description

**Comment by mohammedzee1000** [comment]
Template: https://steps.svc.ci.openshift.org/help/ci-operator#periodic

**Comment by amitkrout** [comment]
Task need to be done

- [ ] Separate integration test https://github.com/openshift/odo/pull/3066
- [ ] Create periodic job and update regular test name with the new test name on OpenShift CI https://github.com/openshift/release/pull/8707

**Comment by kadel** [comment]
How are we going to handle the situations when nightly job fails? 
Is there going to be email notification when that happens? Who is going to be responsible for fixing it?

**Comment by amitkrout** [comment]
> How are we going to handle the situations when nightly job fails? Is there going to be email notification when that happens?

We will follow the regular process like we have been doing, file the issue on master and fix it. BTW this is a duration gap of 6hrs periodic job, i mean in every 6 hrs the test (integration+e2e) job will be triggered. Ideally there should be an option of email notification in the job configuration. I need to figure out the way of only getting the e mail notification when the job fails otherwise no email notification is needed. Till then we need to consistently monitor the test results at - https://prow.svc.ci.openshift.org/?repo=openshift%2Fodo  

> Who is going to be responsible for fixing it?

Once the issue is figured out, the package owner will be the first choice for fixing the issue or anybody can volunteer and fix it.

**Comment by kadel** [comment]
> EDIT: As per my slack communication with platform team - OpenShift CI does not support email notifications but they **do support slack alerts** for now. We can opt for slack alerts then.
> 
> >

It would be great if we can enable slack notification in a separate channel in  public odo slack

**Comment by kadel** [comment]
What jobs will be executed as periodic and what jobs will run as PR tests?

**Comment by amitkrout** [comment]
> It would be great if we can enable slack notification in a separate channel in public odo slack

As an experiment i was sending the notification to any random odo channel. I will update the pr to start sending notification to odo public channel.

**Comment by amitkrout** [comment]
> What jobs will be executed as periodic and what jobs will run as PR tests?

**What is a periodic job -** It runs in every 6 hrs on master branch. e2e and integration test will be the part of thsi run.
**NOTE:** Periodic time can be updated

**Which test will be run against each pr then -**  To optimize the CI total run time we can run only integration test against each pr. e2e test will be run as a periodic test on odo master along with integration test to add more test stability on master.

**Comment by kadel** [comment]
> **What is a periodic job -** It runs in every 6 hrs on master branch. e2e and integration test will be the part of thsi run.
> **NOTE:** Periodic time can be updated

Wouldn't 12 hours be enough? 6 hours feels quite fast, especially as most of the time there will be no new PR merged in 6 hours time window.



> **Which test will be run against each pr then -** To optimize the CI total run time we can run only unit and integration test against each pr. e2e test will be run as a periodic test on odo master along with integration test to add more test stability on master.

How much time we will save by this?


My main concern would be about what benefits we will get running all integration tests for each PR and also as periodic jobs. I was under the impression that thanks to the post-submit tests we should be pretty safe and don't break master.

**Comment by mohammedzee1000** [comment]
It will be a good idea to document the link to periodic tests, what they are, how to update them if needed and how to access them (their status, logs etc)

**Comment by amitkrout** [comment]
> > **What is a periodic job -** It runs in every 6 hrs on master branch. e2e and integration test will be the part of thsi run.
> > **NOTE:** Periodic time can be updated
> 
> Wouldn't 12 hours be enough? 6 hours feels quite fast, especially as most of the time there will be no new PR merged in 6 hours time window.
> 

12 hrs would be ok. Unnecessarily running the test on master does not make sense if there is no new change available on master. However i keep the periodic windows of 6 hrs because to find out the flake and fix those before hand which mostly hurts our pr before merge. You can get repeated failure logs in a quick interval for debugging purpose. Less time interval helps us to run quick experiment like 4 node testing to find out the failure (test/component/devfile...). Atleast we should keep this small windows till we fix the flakes, anyways there will be a revisit time to change periodic window once our test gets more stable.

> > **Which test will be run against each pr then -** To optimize the CI total run time we can run only unit and integration test against each pr. e2e test will be run as a periodic test on odo master along with integration test to add more test stability on master.
> 
> How much time we will save by this?
> 

Total time minus the time taken by e2e test, approximately 20 min of total CI time will be reduced. 

> My main concern would be about what benefits we will get running all integration tests for each PR and also as periodic jobs. I was under the impression that thanks to the post-submit tests we should be pretty safe and don't break master.

The test type we have been using is [Pre-submit](https://ci-operator-configresolver-ui-ci.apps.ci.l2s4.p1.openshiftapps.com/help/ci-operator#test-types) test type since beginning. So to add more test stability on master, e2e test along with integration will also be the part of periodic run. Moreover we really do not bother about the CI time, so it will be helpful to find out the flake that can occurs more frequently on master branch before the pr suffers.

**Comment by kadel** [comment]
> You can get repeated failure logs in a quick interval for debugging purpose. Less time interval helps us to run quick experiment like 4 node testing to find out the failure (test/component/devfile...). Atleast we should keep this small windows till we fix the flakes, anyways there will be a revisit time to change periodic window once our test gets more stable.

It is not possible to trigger tests manually?

**Comment by girishramnani** [comment]
no as per discussion thats not possible to trigger the periodic tests manually.

**Comment by amitkrout** [comment]
> no as per discussion thats not possible to trigger the periodic tests manually.

Yes we can do that just by adding user name to https://github.com/openshift/release/blob/master/core-services/prow/02_config/_config.yaml#L332-L342. For example 
```
'openshift/odo':
      github_users:
      - user1
      - user2
```

**Comment by amitkrout** [comment]
> > EDIT: As per my slack communication with platform team - OpenShift CI does not support email notifications but they **do support slack alerts** for now. We can opt for slack alerts then.
> > >
> 
> It would be great if we can enable slack notification in a separate channel in public odo slack

@kadel Currently our public odo slack is under a free plan i guess which only allow access to few default app for adding it in openshiftdo workspace. If we want to see the job notification then we need to move to a paid plan atleast to get the prow app in the workspace. 

<img width="769" alt="Screen Shot 2020-06-03 at 11 13 15 AM" src="https://user-images.githubusercontent.com/19851115/83601285-e8078480-a58d-11ea-85ee-cb3d86ddef6e.png">

If we are not looking for any paid plan then we have the private channel having prow app installed on that.

**Comment by amitkrout** [comment]
BTW i have asked in slack help center for more details like in free plan how to integrate prow app - https://openshiftdo.slack.com/help/requests/2914186?replied=1#reply

**Comment by amitkrout** [comment]
```
Unfortunately, in order to connect another service with Slack, there needs to be a Slack API app that integrates the two services. Typically, these apps are built by the third-party providers rather than Slack as they know what access they want to allow into their systems.

Therefore, if you want an integration between Slack and the Prow third party service, then you'll want to reach out to the makers of that service to see if they're interested in building a Slack app to provide the functionality that you're seeking. If they are, they can get started on the following page:

https://api.slack.com/slack-apps

That site will walk them through the various integration points we have available, along with details about the requirements for submitting their app for inclusion in our App Directory (https://slack.com/apps).

Also, in case it might help, here's a cheat sheet for getting started with Slack’s APIs:

https://medium.com/slack-developer-blog/getting-started-with-slacks-apis-f930c73fc889

I'm sorry that I don't have better news for you on this, but I hope that the makers of this Prow service are receptive to this idea as there may be other users who want to access their services through Slack.
```

This is the message i got from slack team. It clear that platform team need to allow their prow app to integrate in our public slack ```OpenShift Do``` workspace - https://openshiftdo.slack.com/. Ping @petr-muller @droslean

**Comment by amitkrout** [comment]
> Task need to be done
> 
> * [x]  Separate integration test #3066
> * [x]  Create periodic job on OpenShift CI [openshift/release#9427](https://github.com/openshift/release/pull/9427)
> * [ ]  Email notification if job pass or fails - No support yet, but it supports slack notification
> * [ ]  Increase test node to 4 (Experiment purpose - To find out the flake in 4 test node)

As per the 1-1 discussion with @kadel, we just made a little change in the plan to avoid any risk at the beginning. The plan is to go step by step.

Updated plan:

1. Our pr test will run as it is
2. A periodic job will be added 
3. For now we will not sending a slack notification (We will check manually from the prow dashboard may be in each morning)
4. Will start fixing test flake to stabilize odo test.
5. Then start separating out the test pr* (Integration) and nightly* (Integration + e2e)
6. Configure notification 

Note: ```*``` is for negotiation.

**Comment by kadel** [comment]
> 3\. For now we will not sending a slack notification (We will check manually from the prow dashboard may be in each morning)

can you add a link to the dashboard?

**Comment by amitkrout** [comment]
> > 3. For now we will not sending a slack notification (We will check manually from the prow dashboard may be in each morning)
> 
> can you add a link to the dashboard?

https://prow.svc.ci.openshift.org/?repo=openshift%2Fodo

**Comment by amitkrout** [comment]
> > > 1. For now we will not sending a slack notification (We will check manually from the prow dashboard may be in each morning)
> > 
> > 
> > can you add a link to the dashboard?
> 
> https://prow.svc.ci.openshift.org/?repo=openshift%2Fodo

Periodic jobs will be listed here - https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/?type=periodic&job=periodic-*odo*

**Comment by dharmit** [comment]
Moved this to In Progres as there's nothing really to review. Amit's working on the final bits about sending notifications to slacks.

**Comment by girishramnani** [comment]
adding a link in the docs and notification to slack channel is left

**Comment by kadel** [comment]
> adding a link in the docs and notification to slack channel is left

Where we are on that? Is this getting blocked by something?

**Comment by amitkrout** [comment]
> > adding a link in the docs and notification to slack channel is left
> 
> Where we are on that? Is this getting blocked by something?

@kadel Yes we are blocked by https://github.com/kubernetes/test-infra/issues/18190. Prow does support notification on multiple slack namespace. Currently we have prow available for coreOS slack namespace, so the notification feature will not be supported for https://openshiftdo.slack.com namespace. Anyway we can track the feature request in the issue  https://github.com/kubernetes/test-infra/issues/18190

As part of adding the link is concerned i can open up at any moment.

**Comment by kadel** [comment]
> @kadel Yes we are blocked by [kubernetes/test-infra#18190](https://github.com/kubernetes/test-infra/issues/18190). Prow does support notification on multiple slack namespace from a single instance of prow. Currently we have prow available for coreOS slack namespace, so the notification feature will not be supported for https://openshiftdo.slack.com namespace. Anyway we can track the feature request in the issue [kubernetes/test-infra#18190](https://github.com/kubernetes/test-infra/issues/18190)

Can we at least setup notification for our channel in  coreos slack?

**Comment by amitkrout** [comment]
> > @kadel Yes we are blocked by [kubernetes/test-infra#18190](https://github.com/kubernetes/test-infra/issues/18190). Prow does support notification on multiple slack namespace from a single instance of prow. Currently we have prow available for coreOS slack namespace, so the notification feature will not be supported for https://openshiftdo.slack.com namespace. Anyway we can track the feature request in the issue [kubernetes/test-infra#18190](https://github.com/kubernetes/test-infra/issues/18190)
> 
> Can we at least setup notification for our channel in coreos slack?
> Without notifications the periodic jobs are useless :-(

Yes, this is doable. I will send a pr.

**Comment by amitkrout** [comment]
I have done the changes through the pr https://github.com/openshift/release/pull/10191 to send the periodic job notification to channel ```odo-notifications``` that i configured with.
```
reporter_config:
       slack:
               channel: '#odo-notifications'
```
but did not work as expected. No notification sent to ```#odo-notifications``` along with all periodic job fails with error - https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/10191/rehearse-10191-periodic-ci-openshift-odo-master-v4.2-integration-e2e-periodics/1288013745720659968#1:build-log.txt%3A7

I can not figure out the potential fix from the error statement. I asked it on ```#forum-testplatform``` for a resolution.

**Comment by kadel** [comment]
This is not done.
The test is still falling, there not a single run where would all periodic jobs succeded.

**Comment by amitkrout** [comment]
> This is not done.
> The test is still falling, there not a single run where would all periodic jobs succeded.

Yes, its due to platform issue and known flakes. One major periodic blocker blocker is being worked on pr https://github.com/openshift/odo/pull/3737.

**Comment by amitkrout** [comment]
> This is not done.
 The test is still falling, there not a single run where would all periodic jobs succeded.

The jobs are failing due to platform issue or flakes. However we have few successful job run

<img width="694" alt="Screen Shot 2020-08-21 at 1 34 59 AM" src="https://user-images.githubusercontent.com/19851115/90820154-91207880-e34e-11ea-9fde-344d26dddfbe.png">

**Comment by kadel** [comment]
Are we tracking those issues somewhere?
There was not a day where would all periodic jobs were successful. 

In the current state, periodic jobs are of no help :(
